doc/dalitool.1
.TH DALITOOL 1 2013/10/04/
.SH NAME
dalitool \- DataLink client for data stream inspection and server queries

.SH SYNOPSIS
.nf
dalitool [options] [host][:][port] [repeat]

.fi
.SH DESCRIPTION
\fBdalitool\fP connects to a \fIDataLink\fR server and either queries
the server for information or collects selected data streams and prints
information about the packets received.  All received packets can
optionally be written to a single file.

\fBdalitool\fP also has an interactive command-line console mode where
DataLink commands are sent to the server and the response is printed.
Console mode is initiated with the \fB-c\fP option.

Results of server information queries can either be printed as raw XML
as returned by the server or as formatted output.  If a repeat
interval is specified with an information query the query will be
repeated at the given interval in seconds.

.SH OPTIONS

.IP "-V         "
Report program version and exit.

.IP "-h         "
Print program usage and exit.

.IP "-v         "
Be more verbose.  This flag can be used multiple times ("-v -v" or
"-vv") for more verbosity.

.IP "-c         "
Connect to specified DataLink server and enter console mode, an
interactive command and response prompt.  Type 'help' to list available
commands.

.IP "-p         "
Print details of received packets.  This flag can be used multiple
times ("-p -p" or "-pp") for more detail.

.IP "-d         "
Print the first 6 sample values for each received time-series packet,
implies at least one -p flag.

.IP "-D         "
Print all sample values for each received time-series packet, implies
at least one -p flag.

.IP "-m \fImatch\fR"
Specify a matching expression to send to the server.  This regular
expression is used to either limit the stream packets collected or the
streams or clients reported in information queries.  For packets this
expression is matched against the stream ID, nominally in the
form 'NET_STA_LOC_CHAN/TYPE'.  For clients this expression is matched
against the client address (host:port) string and the client ID, a
successful match against either means the client will be included in
the results.  If the expression begins with an '@' character it is
assumed to be a file containing a list of expressions for matching.

.IP "-r \fIreject\fR"
Specify a rejecting expression to send to the server.  This regular
expression is used to limit the stream packets collected and is
logically opposite of the matching expression.  This expression is
matched against the stream ID, nominally in the
form 'NET_STA_LOC_CHAN/TYPE'.  If the expression begins with an '@'
character it is assumed to be a file containing a list of expressions
for rejecting.

.IP "-k \fIinterval\fR"
Specify keepalive packet interval (in seconds) at which keepalive
packets are sent to the server.  Keepalive packets are only sent if
nothing is received within the interval.

.IP "-x \fIstatefile\fR"
During client shutdown the last received packet ID and packet creation
time will be saved in this file.  If this file exists upon startup the
information will be used to resume the data collection from the point
at which it was stopped.  In this way the client can be stopped and
started without data loss, assuming the data are still available on
the server.

.IP "-o \fIoutfile\fR"
If specified, all received packets will be appended to this file.  The
file is created if it does not exist.  A special mode for this option
is to send all received packets to standard output when the outfile is
specified as '-'.  In this case all diagnostic program output will be
redirected to standard error.

.IP "-i \fItype\fR"
Send an information request; the returned raw XML response is printed.
Supported information types are: STATUS, STREAMS and CONNECTIONS.  The
results of STREAMS and CONNNECTIONS queries can be limited to specific
streams or connections using the \fB-m\fP and \fB-r\fP options.
.PP
Formatted information requests can be made with these options:

.nf
-I   print server status and either exit or repeat at interval
-Q   print stream list and either exit or repeat at interval
-C   print connection list and either exit or repeat at interval
.fi

.IP "-f         "
Increase the level of detail printed for formatted INFO responses.
This flag can be used multiple times ("-f -f" or "-ff") for increased
levels.

.IP "\fI[host][:][port]\fR"
A required argument, specifies the address of the DataLink server in
host:port format.  Either the host, port or both can be omitted.  If
host is omitted then localhost is assumed, i.e. ':16000'
implies 'localhost:16000'.  If the port is omitted then 16000 is
assumed, i.e. 'localhost' implies 'localhost:16000'.  If only ':' is
specified 'localhost:16000' is assumed.

.IP "\fI[repeat]\fR"
A repeat interval in seconds for server information queries.

.SH "EXAMPLES"
.IP
The following would connect to a DataLink server at data.host.edu port
16000 and print details of all received packets.  Since no match or
reject patterns were specified the server will send all data.

.B >dalitool -p data.host.edu:16000

The '-m' and '-r' arguments can be used to limit the data packets, in
the following packets with stream IDs beginning with "US" are matched
and packets with stream IDs beginning with "US_DMGT" are rejected.
This would effectively select the US network data streams with the
exception of station DGMT.  Port 16000 will be assumed as none is
specified.

.B >dalitool -v -p -m '^US.*' -r '^US_DGMT.*' data.host.edu

.IP
The following requests and formats a list of current streams in the
server.

.B >dalitool -S data.host.edu

An information request can be continuously repeated by supplying a
repeat interval, for example to print the connection list every 5
seconds:

.B >dalitool -C data.host.edu 5

.SH AUTHOR
.nf
Chad Trabant
EarthScope Data Services
.fi


ChangeLog
2023.335:
	- Update libdali to 1.8.1
	- Update libmseed to 3.0.17

2020.109:
	- Update libmseed to 2.19.6.
	- Update libdali to 1.8.
	- fix compiler warnings.

2013.280:
	- Add include of time.h in dlconsole.c, needed for Linux, etc.

2013.277:
	- Update libmseed to 2.12.

2013.230:
	- Console mode: check connection status before each command and 
	re-connect if connection is broken.
	- Add CONNECT command to console mode.

2013.223:
	- Update libdali to 1.6.
	- Update libmseed to 2.11.
	- Add console mode (-c option) with line editing/history.  Allows
	interactive DataLink command and response.

2011.262:
	- Update libmseed to 2.6.1.
	- Update libdali to 1.5.

2011.003:
	- Update libdali to 1.3.

2010.027:
	- Update libdali to 1.2.
	- Include printing of ScanTime as returned in STATUS info requests.
	- Cleanup STATUS info output for clarity.

2010.020:
	- Remove parsing of non-portable thread ID in Thread element of 
	STATUS return information.

2010.018:
	- Add verbose printing of new ServerThreads element returned for
	a STATUS (ID) request.
	- Add verbose printing of new StreamCount attribute returned for
	a CONNECTIONS request.

2009.292:
	- Change the formatted STREAMS output to report the DataLatency
	returned by the server instead of a client-side calculation.

2008.256:
	- Update libdali to 1.1.
	- Rework printing of packet due to availability of packet
	data end time in the packet headers.
	- Change the formatted STREAMS output to calculate a stream's
	latency as the difference between the current time and the
	end time of the stream, now equal to DMC's "data latency".
	- Add the -f option to increase the level of details printed
	when formatting INFO responses.
	- Add volatile ring status to formatted STATUS output.

2008.231:
	- Include distinct IP and hostnames as reported by the 
	server to the formatted client list output.
	- Remove hostname resolution code.

2008.218:
	- Add resolved hostnames to formatted client list output.

2008.196:
	- Update to libdali 1.0.
	- Include more details in formatted Status INFO requests.

2008.080:
	- Update to libdali 0.9.4.
	- Send INFO CONNECTIONS match expression regardless of case.
	- More server status for verbose formatted CONNECTIONS output.
	
2008.072:
	- Update to libdali 0.9.2.

2008.070:
	- Update to libdali 0.9.1.
	- For INFO CONNECTIONS requests submit a client matching pattern
	if -m is specified.
	- Include both total and selected counts of connections and
	streams in formatted INFO output.

2008.068:
	- Finish packet handler routine, add data feed latency.
	- Add repeat INFO request logic.
	- Complete INFO STREAMS and CONNECTIONS output.
	- Fixup INFO STATUS output.

2008.066:
	- Initial version.


src/dalitool.c
/***************************************************************************
 *
 * A DataLink client for data stream inspection and server information.
 *
 * Written by Chad Trabant
 *   EarthScope Data Services
 ***************************************************************************/

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#ifndef WIN32
#include <signal.h>
#endif

#include <libdali.h>
#include <libmseed.h>

#include "common.h"
#include "dalixml.h"
#include "dlconsole.h"

#define PACKAGE "dalitool"
#define VERSION "2023.335"

static char verbose        = 0; /* Flag to control general verbosity */
static char console        = 0; /* Flag to control interactive console session */
static char ppackets       = 0; /* Flag to control printing of data packets */
static char psamples       = 0; /* Flag to control printing of data samples */
static char formatinfo     = 0; /* Flag to control formatting of INFO XML */
static int repeatint       = 0; /* Repeat interval for INFO requests */
static char formatlevel    = 0; /* Flag to control formatted output verbosity */
static char *statefile     = 0; /* State file for saving/restoring the seq. no. */
static char *matchpattern  = 0; /* Source ID matching expression */
static char *rejectpattern = 0; /* Source ID rejecting expression */
static char *clientpattern = 0; /* Client matching expression */
static char *infotype      = 0; /* INFO type to request */
static char *outfile       = 0; /* The output file */
static FILE *outfp         = 0; /* Output file descriptor */

static DLCP *dlconn; /* connection parameters */

/* Functions internal to this source file */
static int parameter_proc (int argcount, char **argvec);
static char *getoptval (int argcount, char **argvec, int argopt);
static void print_stderr (const char *message);
static void usage (void);

#ifndef WIN32
static void term_handler (int sig);
#endif

int
main (int argc, char **argv)
{
  DLPacket dlpacket;
  char packetdata[MAXPACKETSIZE];
  char *infobuf = 0;
  int infolen;

  dltime_t current;
  dltime_t next;

#ifndef WIN32
  /* Signal handling, use POSIX calls with standardized semantics */
  struct sigaction sa;

  sa.sa_flags = SA_RESTART;
  sigemptyset (&sa.sa_mask);

  sa.sa_handler = term_handler;
  sigaction (SIGINT, &sa, NULL);
  sigaction (SIGQUIT, &sa, NULL);
  sigaction (SIGTERM, &sa, NULL);

  sa.sa_handler = SIG_IGN;
  sigaction (SIGHUP, &sa, NULL);
  sigaction (SIGPIPE, &sa, NULL);
#endif

  /* Process given parameters (command line and parameter file) */
  if (parameter_proc (argc, argv) < 0)
  {
    dl_log (2, 0, "Parameter processing failed\n\n");
    dl_log (2, 0, "Try '-h' for detailed help\n");
    return -1;
  }

  /* Connect to server */
  if (dl_connect (dlconn) < 0)
  {
    dl_log (2, 0, "Error connecting to server\n");
    return -1;
  }

  /* Reposition connection */
  if (dlconn->pktid > 0)
  {
    if (dl_position (dlconn, dlconn->pktid, dlconn->pkttime) < 0)
      return -1;
  }

  /* Send match pattern if supplied */
  if (matchpattern)
  {
    if (dl_match (dlconn, matchpattern) < 0)
      return -1;
  }

  /* Send reject pattern if supplied */
  if (rejectpattern)
  {
    if (dl_reject (dlconn, rejectpattern) < 0)
      return -1;
  }

  /* Request INFO and print returned XML either formatted nicely or not */
  if (infotype)
  {
    /* Initialize next request time stamp */
    next = dlp_time ();

    while (!dlconn->terminate)
    {
      current = dlp_time ();

      /* Submit INFO request and process response if at the next time stamp */
      if (current >= next)
      {
        if ((infolen = dl_getinfo (dlconn, infotype, clientpattern, &infobuf, 0)) < 0)
        {
          dl_log (2, 0, "Problem requesting INFO from server\n");
          return -1;
        }

        /* Print formatted INFO */
        if (formatinfo)
          info_handler (infotype, infobuf, infolen, formatlevel);
        /* Print raw XML */
        else
          printf ("%.*s\n", infolen, infobuf);

        if (infobuf)
          free (infobuf);

        /* Update the next time stamp to the repeat interval */
        next += (repeatint * DLTMODULUS);

        /* If the next time stamp is not in the future move up */
        if (next < current)
          next = current + (repeatint * DLTMODULUS);

        /* If repeating add a newline to separate output */
        if (repeatint)
          printf ("\n");
      }
      /* Otherwise sleep for 1/10 second */
      else
      {
        dlp_usleep (100000);
      }

      /* If not repeating jump out */
      if (!repeatint)
        break;
    }
  }
  /* Enter interactive console mode */
  else if (console)
  {
    if (runconsole (dlconn, &dlpacket, packetdata, sizeof (packetdata), verbose))
    {
      dl_log (2, 0, "Error running console()\n");
    }
  }
  /* Otherwise collect packets in STREAMing mode */
  else
  {
    /* Collect packets in streaming mode */
    while (dl_collect (dlconn, &dlpacket, packetdata, sizeof (packetdata), 0) == DLPACKET)
    {
      packet_handler (&dlpacket, packetdata, ppackets, psamples, (outfile && outfp) ? outfp : NULL);
    }
  }

  /* Shutdown */
  if (dlconn->link != -1)
    dl_disconnect (dlconn);

  if (outfp)
    fclose (outfp);

  if (statefile)
    dl_savestate (dlconn, statefile);

  return 0;
} /* End of main() */

/***************************************************************************
 * parameter_proc:
 *
 * Process the command line parameters.
 *
 * Returns 0 on success, and -1 on failure
 ***************************************************************************/
static int
parameter_proc (int argcount, char **argvec)
{
  char *address = 0;
  int keepalive = -1;
  int optind;

  /* Process all command line arguments */
  for (optind = 1; optind < argcount; optind++)
  {
    if (strcmp (argvec[optind], "-V") == 0)
    {
      fprintf (stderr, "%s version: %s\n", PACKAGE, VERSION);
      exit (0);
    }
    else if (strcmp (argvec[optind], "-h") == 0)
    {
      usage ();
      exit (0);
    }
    else if (strncmp (argvec[optind], "-v", 2) == 0)
    {
      verbose += strspn (&argvec[optind][1], "v");
    }
    else if (strncmp (argvec[optind], "-c", 2) == 0)
    {
      console = 1;
    }
    else if (strncmp (argvec[optind], "-p", 2) == 0)
    {
      ppackets += strspn (&argvec[optind][1], "p");
    }
    else if (strcmp (argvec[optind], "-d") == 0)
    {
      psamples = 1;
    }
    else if (strcmp (argvec[optind], "-D") == 0)
    {
      psamples = 2;
    }
    else if (strcmp (argvec[optind], "-k") == 0)
    {
      keepalive = strtoul (getoptval (argcount, argvec, optind++), NULL, 10);
    }
    else if (strcmp (argvec[optind], "-o") == 0)
    {
      outfile = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-x") == 0)
    {
      statefile = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-m") == 0)
    {
      matchpattern = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-r") == 0)
    {
      rejectpattern = getoptval (argcount, argvec, optind++);
    }
    else if (strcmp (argvec[optind], "-i") == 0)
    {
      infotype   = getoptval (argcount, argvec, optind++);
      formatinfo = 0;
    }
    else if (strcmp (argvec[optind], "-I") == 0)
    {
      infotype   = "STATUS";
      formatinfo = 1;
    }
    else if (strcmp (argvec[optind], "-S") == 0)
    {
      infotype   = "STREAMS";
      formatinfo = 1;
    }
    else if (strcmp (argvec[optind], "-C") == 0)
    {
      infotype   = "CONNECTIONS";
      formatinfo = 1;
    }
    else if (strncmp (argvec[optind], "-f", 2) == 0)
    {
      formatlevel += strspn (&argvec[optind][1], "f");
    }
    else if (strncmp (argvec[optind], "-", 1) == 0)
    {
      fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
      exit (1);
    }
    else if (!address)
    {
      address = argvec[optind];
    }
    else if (!repeatint)
    {
      repeatint = strtoul (argvec[optind], NULL, 10);
    }
    else
    {
      fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
      exit (1);
    }
  }

  /* Make sure a server was specified */
  if (!address)
  {
    fprintf (stderr, "No DataLink server specified\n\n");
    fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
    fprintf (stderr, "Usage: %s [options] [host][:][port] [int]\n\n", PACKAGE);
    fprintf (stderr, "Try '-h' for detailed help\n");
    exit (1);
  }

  /* Allocate and initialize a new connection description */
  dlconn = dl_newdlcp (address, argvec[0]);

  /* Set keepalive parameter, allow for valid value of 0 */
  if (keepalive >= 0)
    dlconn->keepalive = keepalive;

  /* Initialize the verbosity for the dl_log function */
  dl_loginit (verbose, NULL, NULL, NULL, NULL);

  /* Open output file if requested */
  if (outfile)
  {
    if (!strcmp (outfile, "-"))
    {
      /* Re-direct all messages to standard error */
      dl_loginit (verbose, &print_stderr, NULL, &print_stderr, NULL);

      outfp = stdout;
      setvbuf (stdout, NULL, _IONBF, 0);
    }
    else if ((outfp = fopen (outfile, "a+b")) != NULL)
    {
      setvbuf (outfp, NULL, _IONBF, 0);
    }
    else
    {
      dl_log (2, 0, "cannot open output file: %s\n", outfile);
      exit (1);
    }
  }

  /* Report the program version */
  dl_log (1, 1, "%s version: %s\n", PACKAGE, VERSION);

  /* Make sure we print basic packet details if printing samples */
  if (psamples && ppackets == 0)
    ppackets = 1;

  /* Load the match stream list from a file if the argument starts with '@' */
  if (matchpattern && *matchpattern == '@')
  {
    char *filename = matchpattern + 1;

    if (!(matchpattern = dl_read_streamlist (dlconn, filename)))
    {
      dl_log (2, 0, "Cannot read matching list file: %s\n", filename);
      exit (1);
    }
  }

  /* Load the reject stream list from a file if the argument starts with '@' */
  if (rejectpattern && *rejectpattern == '@')
  {
    char *filename = rejectpattern + 1;

    if (!(rejectpattern = dl_read_streamlist (dlconn, filename)))
    {
      dl_log (2, 0, "Cannot read rejecting list file: %s\n", filename);
      exit (1);
    }
  }

  /* Set up for INFO CONNECTIONS request and matching */
  if (matchpattern && infotype && !strncasecmp (infotype, "CONNECTIONS", 11))
  {
    clientpattern = matchpattern;
    matchpattern  = 0;
  }

  /* Recover from the state file and reposition */
  if (statefile)
  {
    if (dl_recoverstate (dlconn, statefile) < 0)
    {
      dl_log (2, 0, "Error reading state file\n");
      exit (1);
    }
  }

  return 0;
} /* End of parameter_proc() */

/***************************************************************************
 * getoptval:
 *
 * Return the value to a command line option; checking that the value is
 * itself not an option (starting with '-') and is not past the end of
 * the argument list.
 *
 * argcount: total arguments in argvec
 * argvec: argument list
 * argopt: index of option to process, value is expected to be at argopt+1
 *
 * Returns value on success and exits with error message on failure
 ***************************************************************************/
static char *
getoptval (int argcount, char **argvec, int argopt)
{
  if (argvec == NULL || argvec[argopt] == NULL)
  {
    fprintf (stderr, "getoptval(): NULL option requested\n");
    exit (1);
  }

  /* Special case of '-o -' usage */
  if ((argopt + 1) < argcount && strcmp (argvec[argopt], "-o") == 0)
    if (strcmp (argvec[argopt + 1], "-") == 0)
      return argvec[argopt + 1];

  if ((argopt + 1) < argcount && *argvec[argopt + 1] != '-')
    return argvec[argopt + 1];

  fprintf (stderr, "Option %s requires a value\n", argvec[argopt]);
  exit (1);

  return NULL; /* To stop compiler warnings about no return */
} /* End of getoptval() */

/***************************************************************************
 * print_stderr:
 *
 * Print the given message to standard error.
 ***************************************************************************/
static void
print_stderr (const char *message)
{
  fprintf (stderr, "%s", message);
  return;
}

#ifndef WIN32
/***************************************************************************
 * term_handler:
 *
 * Signal handler routine to set the termination flag.
 ***************************************************************************/
static void
term_handler (int sig)
{
  dl_terminate (dlconn);
}
#endif

/***************************************************************************
 * usage:
 *
 * Print the usage message and exit.
 ***************************************************************************/
static void
usage (void)
{
  fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
  fprintf (stderr, "Usage: %s [options] [host][:][port] [repeat]\n\n", PACKAGE);
  fprintf (stderr,
           " ## General program options ##\n"
           " -V              report program version\n"
           " -h              show this usage message\n"
           " -v              be more verbose, multiple flags can be used\n"
           " -c              console mode, provide an interactive prompt\n"
           " -p              print details of data packets, multiple flags can be used\n"
           " -d              print first 6 samples of each data packet\n"
           " -D              print all samples of each data packet\n"
           " -m match        specify stream ID matching pattern\n"
           " -r reject       specify stream ID rejecting pattern\n"
           " -k interval     send keepalive packets this often (seconds)\n"
           " -x sfile        save/restore state information to this file\n"
           " -o outfile      write all received packets to this file\n"
           "\n"
           " ## Data server information ##\n"
           " -i type         send info request, type is one of the following:\n"
           "                   STATUS, STREAMS, CONNECTIONS\n"
           "                   the returned raw XML is displayed when using this option\n"
           " -I              print formatted server id, version and status\n"
           " -S              print formatted stream list (if supported by server)\n"
           " -C              print formatted connection list (if supported by server)\n"
           " -f              increase level of details included in formatted output\n"
           "\n"
           " [host][:][port] Address of the DataLink server in host:port format\n"
           "                   Default host is 'localhost' and default port is '16000'\n"
           "\n"
           " [repeat]        Specify a repeat interval in seconds for INFO requests\n"
           "\n");

} /* End of usage() */
